Find nested tags

alecgibson opened the pull request:
This rake task will process all content tagged under a root taxon, and find any content tagged to a taxon where the content is also tagged to an ancestor of that taxon.

### Trello

https://trello.com/c/cuRxpiUD/465-create-a-list-of-all-content-pages-tagged-both-at-a-leaf-taxon-and-at-another-taxon-up-in-that-same-tree

suzannehamilton commented on lib/tasks/taxonomy/find_nested_tags.rake:
```diff
@@ -0,0 +1,55 @@
+      if content_already_tagged.key?(content_base_path)
+        errors << "#{content_base_path} tagged at #{taxon_base_path}, but also tagged at ancestors: #{content_already_tagged[content_base_path].to_a}"
+      else
+        content_already_tagged[content_base_path] = Set.new
```
You can avoid doing this by initializing the hash using `Hash.new { |h, k| h[k] = Set.new }` rather than `{}`.

alecgibson replied:
I'm not sure I understand?

alecgibson replied:
I recognise that this would be nice, but the niceness isn't as nice with the new clone method I need.

suzannehamilton commented on lib/tasks/taxonomy/find_nested_tags.rake:
```diff
@@ -0,0 +1,55 @@
+
+    root_taxon = content_store.content_item(root_taxon_base_path).to_h
+
+    errors = find_content_tagged_at_multiple_levels(root_taxon)
```
I wonder if there's a better name than "errors". There's nothing in the task's name or description which hints that content tagged at multiple levels is an error, and when I first saw it I assumed it was rake task error handling rather than the expected result.

alecgibson replied:
Done

carvil commented on lib/tasks/taxonomy/find_nested_tags.rake:
```diff
@@ -0,0 +1,55 @@
+
+    content_store = GdsApi::ContentStore.new(Plek.find('content-store'))
+
+    root_taxon = content_store.content_item(root_taxon_base_path).to_h
```
You should probably get it from the publishing API to avoid another dependency, as this is a publishing app. That means either passing in a content_id, or do a search by base_path.

An example with content ID would be:

```
Services.publishing_api.get_content(taxon_content_id).to_h
```

alecgibson replied:
Done

suzannehamilton commented on lib/tasks/taxonomy/find_nested_tags.rake:
```diff
@@ -0,0 +1,55 @@
+
+    if child_taxons
+      child_taxons.each do |child_taxon|
+        errors += find_content_tagged_at_multiple_levels(child_taxon, content_already_tagged.clone)
```
Just to check - does `clone` create copies of the sets in the hash?

alecgibson replied:
Good catch - fixed

alecgibson commented:
@suzannehamilton @carvil I've addressed your comments - can you please re-review?

I've run it on int here: https://deploy.integration.publishing.service.gov.uk/job/run-rake-task/1327/console

carvil commented:
@alecgibson could you please output this in a CSV format? It would be nice to copy/paste the errors into a spreadsheet.

Could you please speak with Holly or Graeme to see what the best format would be for this?

alecgibson commented:
@carvil I've updated it to output CSV, but still waiting on Holly and/or Graeme's input.

alecgibson commented:
@carvil @suzannehamilton this is now good to go. It's outputting CSV. After all that, I don't think there are actually any content items with triple-nested taxons (which would be pretty crazy anyway?).

https://deploy.integration.publishing.service.gov.uk/job/run-rake-task/1330/console

alecgibson commented:
Updated to include organisations: https://deploy.integration.publishing.service.gov.uk/job/run-rake-task/1342/console

carvil commented on lib/tasks/taxonomy/find_nested_tags.rake:
```diff
@@ -0,0 +1,124 @@
+
+    nested_tag_warnings = find_content_tagged_at_multiple_levels(taxonomy.child_expansion)
+
+    output = format_warnings_as_rows nested_tag_warnings
```
Could you please wrap arguments in parenthesis? It' makes it easier to read and consistent with other method calls.

alecgibson replied:
Done

alecgibson commented:
@suzannehamilton could you please re-review?
